jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,602 @@
+		checkIfValidLayoutAndType(functionMethodType, functionDescriptor);
+		
+		/* As explained in the Spec of LibraryLookup, the downcall must hold a strong reference to
+		 * the native library symbol to prevent the underlying native library from being unloaded
+		 * during the native calls.
+		 * 
+		 * Note: the passed-in addressable parameter can be either LibraryLookup.Symbol or MemoryAddress.
+		 */
+		functionAddr = downcallAddr;
+		funcMethodType = functionMethodType;
+		funcDescriptor = functionDescriptor;
+		cifNativeThunkAddr = 0;
+		argTypesAddr = 0;
+		generateAdapter();
+	}
+	
+	/* Map the layouts of return type & argument types to the underlying prep_cif */
+	private void generateAdapter() {
+		/* Set the void layout string intended for the underlying native code as the corresponding layout doesn't exist in the Spec */
+		String retLayoutStr = (realReturnLayout == null) ? "b0[abi/kind=VOID]" : realReturnLayout.toString(); //$NON-NLS-1$
+		
+		int argLayoutCount = argLayoutArray.length;
+		String[] argLayoutStrs = new String[argLayoutCount];
+		for (int argIndex = 0; argIndex < argLayoutCount; argIndex++) {
+			MemoryLayout argLayout = argLayoutArray[argIndex];
+			argLayoutStrs[argIndex] = argLayout.toString();
+		}
+		
+		synchronized(privateClassLock) {
+			/* If a prep_cif for a given function descriptor exists, then the corresponding return & argument layouts
+			 * were already set up for this prep_cif, in which case there is no need to check the layouts.
+			 * If not the case, check at first whether the same return & argument layouts exist in the cache
+			 * in case of duplicate memory allocation for the same layouts.
+			 * 
+			 * Note: an int method (e.g. '(int,int)int') and a method function (e.g.'(boolean,boolean)boolean') share
+			 * the same C_INT layouts (e.g. 'b32[abi/kind=INT]b32[abi/kind=INT])b32[abi/kind=INT') but the underlying C
+			 * functions might be different from each other as follows:
+			 * e.g. 
+			 * int f1(int intArg1, int intArg2){ (the layouts are C_INT)
+			 *     int sum = intArg1 + intArg2;
+			 *     return sum;
+			 * } 
+			 * and 
+			 * int f2(int boolArg1, bool boolArg2) (the layouts are C_INT)
+			 * {  
+			 *     int boolSum = boolArg1 || boolArg2;
+			 *     return boolSum;
+			 * }
+			 * 
+			 * To avoid overriding the prep_cif that is already cached for the same C_INT layouts, the hashcode of
+			 * the function descriptor plus the corresponding function address is used to determine whether the int
+			 * function and the boolean function come from the same C function or not; otherwise, a new prep_cif intended
+			 * for the boolean function should be created in the native code.
+			 */
+			long functionAddrLongValue = functionAddr.address().toRawLongValue();
+			String funcDescHashPlusFuncAddr = funcDescriptor.hashCode() + "@" + functionAddrLongValue; //$NON-NLS-1$
```

Comment:
Sorry, I'm still a little confused here. If the address is part of the key, doesn't that mean that no other function can have the same key.

This is my understanding:
```
int func1(int);
int func2(int);
```

funcDescHashPlusFuncAddr = "0x123"+ "@" + "[uniqueAddr]" //func1
funcDescHashPlusFuncAddr = "0x123"+ "@" + "[uniqueAddr]" //func2

So in this case `func1` and `func2` have the same signature, so they should share the same cif, however they will never have the same `addr` so sharing is not possible. Or did I miss something?

Reply:
> So in this case func1 and func2 have the same signature, so they should share the same cif, however they will never have the same addr so sharing is not possible.

Yes. The intent of combing the function address to the cache key is to guarantee the uniqueness of the keys for all function descriptors so as to differentiate `boolean` from `int` with the same `C_INT` layout or `char` from `short` with same `C_SHORT` layout. Otherwise, there is no need to append the function address. 

Technically I am not quite sure whether there is any potential issue in the future to share the `cif` when it comes to different method types with the same layouts. If not, I can simply remove the function address from the key to see whether it still works good as expected.

Reply:
I removed all the function address from the cache key and it seems everything works good as expected.
e.g. the following functions
``` 
int func1(int)
int func2(int)` 
boolean func3(boolean)
```
share the same cif given their have the same `C_INT` layouts. So there should be problem with the the adjustment for now.

We might need to keep any eye on this in case there is any update in the Spec as to how to handle `boolean`.